### PR TITLE
For lastplayed time support

### PR DESCRIPTION
**Pull request by a358003542**
we can get the lastplayed time data from the configparser , but seems there is no way get the total palytime data, so which data in `GameTime` return is temporarily set to None.

### CODE REVIEW

File: src/local_client_base.py [hunk 1]
```diff
@@ -50,13 +50,16 @@ def __init__(self, update_statuses):
...
+        self.load_local_files_done = False
         self.installed_games_cache = self.get_installed_games()
 
         loop = asyncio.get_event_loop()
         loop.create_task(self._register_local_data_watcher())
         loop.create_task(self._register_classic_games_updater())
         self.classic_games_parsing_task = None
 
+
+
```
Comment (bartok765) [inline src/local_client_base.py]: according to pep8 every method should be separated by only one empty line https://www.python.org/dev/peps/pep-0008/#blank-lines

File: src/plugin.py [hunk 2]
```diff
@@ -452,6 +452,22 @@ def _get_not_added_free_games(owned_games):
...
         await self.authentication_client.shutdown()
 
+    async def get_game_time(self, game_id, context):
+        if self.local_client.load_local_files_done:
+            for game in self.local_client.config_parser.games:
+                try:
+                    blizzard_game = Blizzard[game.uid]
+                except KeyError:
+                    log.warning(f'[{game.uid}] is not known blizzard game. Skipping')
+                    return None
```
Comment (bartok765) [inline src/plugin.py]: We should not return `None`, only `GameTime`:
https://galaxy-integrations-python-api.readthedocs.io/en/latest/galaxy.api.html#galaxy.api.plugin.Plugin.get_game_time

What about:
```suggestion
                    continue
```

File: src/plugin.py [hunk 3]
```diff
@@ -452,6 +452,22 @@ def _get_not_added_free_games(owned_games):
...
+
+                if blizzard_game.id == game_id:
+                    last_played = int(game.last_played)
```
Comment (bartok765) [inline src/plugin.py]: game.last_played may be None. This should be covered.

File: src/plugin.py [hunk 4]
```diff
@@ -452,6 +452,22 @@ def _get_not_added_free_games(owned_games):
...
+                    log.debug(f'get_game_time return {game_id} {last_played}')
+                    return GameTime(game_id, None, last_played_time=last_played)
+
+        log.warning(f'warning get_game_time failed for {game_id}')
```
Comment (bartok765) [inline src/plugin.py]: what about:
```python
last_played = None
for game in self.local_client.config_parser.games:
     ...
     if game.last_played is not None:
         last_played = int(game.last_played)
         break

return GameTime(game_id, None, last_played_time=last_played)
```

This would be analogous to #7

File: src/local_client_base.py (hunk 1 above)
Comment (a358003542) [reply src/local_client_base.py]: agreed

File: src/plugin.py (hunk 2 above)
Comment (a358003542) [reply src/plugin.py]: @bartok765 Yes,   should return like that : `GameTime(id, None, None)` . this is will be changed.

File: src/plugin.py (hunk 3 above)
Comment (a358003542) [reply src/plugin.py]: @bartok765 OK

File: src/plugin.py (hunk 4 above)
Comment (a358003542) [reply src/plugin.py]: @bartok765 Yes , need some merge work.

File: src/plugin.py [hunk 5]
```diff
@@ -452,6 +452,22 @@ def _get_not_added_free_games(owned_games):
         log.info("Plugin shutdown.")
         await self.authentication_client.shutdown()
 
+    async def get_game_time(self, game_id, context):
+        if self.local_client.load_local_files_done:
```
[the remark this answers is not in the document]
Comment (a358003542) [reply src/plugin.py]: @bartok765 I will be thinking about it.

**Comment by a358003542** [comment]
@bartok765 sorry that i am know a little about the pytest and mock.

File: tests/conftest.py [hunk 6]
```diff
@@ -75,6 +75,10 @@ def config_data():
         "destiny2": {
             "ServerUid": "destiny2"
         },
+        "prometheus": {
+            "LastPlayed": "1582350061",
+            "LastActioned": "1582208289"
+        },
```
Comment (bartok765) [inline tests/conftest.py]: no need as I commented in above

File: tests/conftest.py [hunk 7]
```diff
@@ -120,8 +124,10 @@ def config_data():
 
 
 @pytest.fixture()
-def local_client_mock():
-    return Mock()
+def local_client_mock(config_data):
+    mock = Mock()
+    mock.config_parser = ConfigParser(config_data)
+    return mock
```
Comment (bartok765) [inline tests/conftest.py]: not need as I commented above. I'll create full local client mock if some tests not related with local_client will need it

File: tests/conftest.py (hunk 6 above)
Comment (a358003542) [reply tests/conftest.py]: ok

File: src/plugin.py [hunk 8]
```diff
@@ -358,11 +359,26 @@ def _get_not_added_free_games(owned_games):
...
-            log.debug(f"Gametime for Overwatch is {game_time_minutes} minutes.")
-        return GameTime(game_id, game_time_minutes, None)
+            total_time = await self._get_overwatch_time()
+            log.debug(f"Gametime for Overwatch is {total_time} minutes.")
+
+        for game in self.local_client.config_parser.games:
+            try:
+                blizzard_game = Blizzard[game.uid]
+            except KeyError:
+                log.warning(f'[{game.uid}] is not known blizzard game. Skipping')
```
[the remark this answers is not in the document]
Comment (a358003542) [reply src/plugin.py]: ok

File: src/plugin.py [hunk 9]
```diff
@@ -358,11 +359,26 @@ def _get_not_added_free_games(owned_games):
...
+                break
```
[the remark this answers is not in the document]
Comment (a358003542) [reply src/plugin.py]: actually in the `Blizzard`  has a searching process, so here break is safe.

[the remark this answers is not in the document]
Comment (a358003542) [reply src/plugin.py]: @bartok765 there is no need to create another new test file , so i add another test function,  the None and the had a data situation all have been tested.

also i need to change the local_client_mock to been a `MagicMock` to make sure it iterable, for the sake of that piece of code: `for game in self.local_client.config_parser.games`

File: tests/test_game_time.py [hunk 10]
```diff
@@ -60,3 +61,14 @@ def ow_player_data_response_not_found():
...
+async def test_last_played_time(pg, config_data):
+    pg.local_client.config_parser = ConfigParser(config_data)
+
+    ctx = await pg.prepare_game_times_context(["21298"])
+    result = await pg.get_game_time("21298", ctx)
+    assert result == GameTime("21298", None, None)
+
+    ctx = await pg.prepare_game_times_context(["1214607983"])
+    result = await pg.get_game_time("1214607983", ctx)
+    assert result == GameTime("1214607983", None, 1441712029)
```
Comment (bartok765) [inline tests/test_game_time.py]: This is fine (though not "perfect")

Best unit-test tests only one thing and does not rely on implementation of other things like ConfigParser which should be tested independently.
But this case is acceptable, because ConfigParser is simple class that does not makes any side-effects like reading files and so on. In case of more complicated things, all of those should be mocked for example:

```python

# conftest

@pytest.fixture()
def config_parser():
    return MagicMock()


@pytest.fixture()
def local_client_mock(config_parser):
    mock = Mock(spec=LocalClient)
    mock.launch_game = AsyncMock()
    mock.config_parser = config_parser
    return mock
```
```python
# test file

from src.definitions import Blizzard, ConfigGameInfo


@pytest.mark.asyncio
async def test_last_played_when_unknown_game_in_config(pg, config_parser):
    game = Blizzard['diablo3']
    unknown_game_uid = 'unknown_game'
    with pytest.raises(KeyError):
        assert Blizzard[unknown_game_uid]  # test precondition

    config_parser.games = [
        ConfigGameInfo(unknown_game_uid, 'mock', '111111111'),
        ConfigGameInfo(game.uid, 'diablo3_enus', '1441712029')
    ]

    ctx = await pg.prepare_game_times_context([game.blizzard_id])
    result = await pg.get_game_time(game.blizzard_id, ctx)
    assert result.game_id == game.blizzard_id
    assert result.last_played_time == 1441712029
```
Test from this example is failing when keeping `break` from previous discussion and is fixed when giving `continue`

But as I said - its OK to have explicite `ConfigParser(mock_data)` here, you don't have to change it

File: tests/test_game_time.py [hunk 11]
```diff
@@ -60,3 +61,14 @@ def ow_player_data_response_not_found():
...
     assert result == GameTime(OW_ID, minutes, None)
+
+async def test_last_played_time(pg, config_data):
+    pg.local_client.config_parser = ConfigParser(config_data)
+
+    ctx = await pg.prepare_game_times_context(["21298"])
+    result = await pg.get_game_time("21298", ctx)
+    assert result == GameTime("21298", None, None)
+
+    ctx = await pg.prepare_game_times_context(["1214607983"])
```
Comment (bartok765) [inline tests/test_game_time.py]: You can prepare just one context with both games - this is the way Galaxy does just before getting game_time for every game. For now does not matter as preparing context is always None, but this maybe changed in the future

```suggestion
    ctx = await pg.prepare_game_times_context(["21298", "1214607983"])
    result = await pg.get_game_time("21298", ctx)
    assert result == GameTime("21298", None, None)

```

File: src/plugin.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (a358003542) [reply src/plugin.py]: @bartok765 first ,we are sure about that the galaxy client call this function with one game id, then use the Blizzard class found the target blizzard_game class. 

second if here program can not found ,  that mean this plugin can not install this game at all , because this plugin has many place is based on the `Blizzard` class load function. so in that siuation which you said the PTR version game, we should add some game definition.py , but not here problem.

File: tests/test_game_time.py (hunk 10 above)
Comment (a358003542) [reply tests/test_game_time.py]: @bartok765 thanks your advice , i should changed that,  thats imporved the test case.

File: tests/test_game_time.py (hunk 11 above)
Comment (a358003542) [reply tests/test_game_time.py]: ok

File: tests/test_game_time.py [hunk 12]
```diff
@@ -60,3 +62,35 @@ def ow_player_data_response_not_found():
...
+@pytest.mark.asyncio
+async def test_last_played_when_unknown_game_in_config(pg, config_parser):
+    game = Blizzard['diablo3']
+    unknown_game_uid = 'unknown_game'
+    with pytest.raises(KeyError):
+        assert Blizzard[unknown_game_uid]  # test precondition
+
+    # this way we fill mocks with desired data we want to test; the trick is pg fixture uses
+    # config_parser internally as part of local_client, but it can be also:
+    # pg.local_client.config_parser.games = ... (but sometimes it's not possible)
```
Comment (bartok765) [inline tests/test_game_time.py]: can be removed. It was just for you to explain

Comment (a358003542) [reply tests/test_game_time.py]: ok done.

**Comment by bartok765** [comment]
Thanks @a358003542 for this!
